stdlib/asm/crypto/stark/deep_queries.masm

```diff
@@ -1,245 +1,471 @@
-#! 6- a_addr is the memory address of the i-th random element used in batching the trace polynomial quotients.
-#! The random elements a := (a0, a1) are stored in memory as [0, 0, a0, a1].
+#! Input: [...]
+#! Output: [p_gz_1, p_gz_0, q_z_1, q_z_0, ...]
 #!
-#! Input: [T31, T30, T21, T20, T11, T10, T01, T00, p1, p0, r1, r0, x_addr, z_addr, a_addr, 0]
-#! Output: [T01, T00, T31, T30, T21, T20, T11, T10, p1', p0', r1', r0', x_addr, z_addr', a_addr', 0]
-export.combine_aux
+#! Cycles: 200
+proc.compute_q_z_p_gz_at_alpha
 
-    # 1) Shift trace columns values (as quadratic extension field element) left
-    movup.7 movup.7
-    #=> [T01, T00, T31, T30, T21, T20, T11, T10, p1, p0, r1, r0, x_addr, z_addr, a_addr, 0]
+    # I) Compute alpha^{-1} is and save it
+    padw exec.constants::deep_rand_coef_ptr mem_loadw
+    # => [y, y, alpha1, alpha0, ...]
+    drop drop
+    ext2inv
+    dup.1 dup.1
+    exec.constants::deep_rand_alpha_inv mem_storew
 
-    # 2) Get a_addr and update it. This is done here before it becomes inaccessible.
+    # II) Compute Q^z(alpha)
 
-    # Update a_addr
-    dup.14 add.4 swap.15
-    #=> [a_addr, T01, T00, T31, T30, T21, T20, T11, T10, p1, p0, r1, r0, x_addr, z_addr, a_addr', 0]
+    ## 1) Set up the stack for `horner_eval_ext`
 
-    # 3) Load i-th OOD frame portion. This assumes that the OOD frame has been serialized with `current` and `next` rows interleaved.
-    # This also updates the z_addr pointer.
-    dup.14 add.4 swap.15
-    padw movup.4 mem_loadw
-    #=> [Tgz1, Tgz0, Tz1, Tz0, a_addr, T01, T00, T31, T30, T21, T20, T11, T10, p1, p0, r1, r0, x_addr, z_addr', a_addr', 0]
+    ### a) Set up the initial accumulator and the pointers to alpha^{-1} and a memory region
+    ###    from which to load the OOD.
+    push.0.0
+    exec.constants::deep_rand_alpha_inv
+    exec.constants::ood_trace_ptr
+    exec.constants::tmp2 mem_storew
 
-    # 4) Compute the numerators
+    ### b) Set up the rest of the stack
+    swapw
+    padw swapw
+    padw
 
-    # a) Compute T_i - T_i(z). This equals, in the case of T0, (T01 - Tz1, T00 - Tz0)
-    dup.6 dup.6
-    movup.5 movup.5
-    #=> [Tz1, Tz0, T01, T00, Tgz1, Tgz0, a_addr, T01, T00, T31, T30, T21, T20, T11, T10, p1, p0, r1, r0, x_addr, z_addr', a_addr', 0]
+    ## 2) Load the execution trace, both main and auxiliary, portion of the OOD evaluation
+    ##    Since (71 + 7) * 2 divmod 8 = (19, 4), we will need to handle the last batch of
+    ##    4 base field elements separately.
+    
+    ### a) Process the first 19 batches of 8 field elements i.e., 4 quadratic extension field elements
+    repeat.19
+        mem_stream
+        horner_eval_ext
+    end
 
-    ext2sub
-    #=> [T01 - Tz1, T00 - Tz0, Tgz1, Tgz0, a_addr, T01, T00, T31, T30, T21, T20, T11, T10, p1, p0, r1, r0, x_addr, z_addr', a_addr', 0]
+    ## 3) Load the last 2 extension field elements. We use `mem_stream` but we will discard the last
+    ##    4 base field elements
+    mem_stream
```

Comment:
instead of hardcoding this `divmod`, could we not use `u32divmod.8 = (q,r)` and treat `q` and `r` "dynamically"? We'd probably pay <100 extra cycles, but it would make the code much more maintainable.

Can we actually afford ~ 100 cycles x however many places we use this pattern (as an upper bound)?

Reply:
The way I am thinking about this is that the verifier, once it stabilizes, it shouldn't change frequently and hence this shouldn't be a big inconvenience. But this is up for discussion for sure.

Reply:
Created https://github.com/0xPolygonMiden/miden-vm/issues/1690